libr/bin/p/bin_mach0.c

```diff
@@ -503,6 +507,179 @@ static RBinInfo *info(RBinFile *bf) {
 	return ret;
 }
 
+static bool _patch_reloc(struct MACH0_(obj_t) *bin, RIOBind *iob, struct reloc_t * reloc, ut64 symbol_at) {
+	ut64 pc = reloc->addr;
+	ut64 ins_len = 0;
+
+	switch (bin->hdr.cputype) {
+	case CPU_TYPE_X86_64: {
+		switch (reloc->type) {
+		case X86_64_RELOC_UNSIGNED:
+			break;
+		case X86_64_RELOC_BRANCH:
+			pc -= 1;
+			ins_len = 5;
+			break;
+		default:
+			eprintf ("Warning: unsupported reloc type for X86_64 (%d), please file a bug.\n", reloc->type);
+			return false;
+		}
+		break;
+	}
+	case CPU_TYPE_ARM64:
+	case CPU_TYPE_ARM64_32:
+		pc = reloc->addr & ~3;
+		ins_len = 4;
+		break;
+	case CPU_TYPE_ARM:
+		break;
+	default:
+		eprintf ("Warning: unsupported architecture for patching relocs, please file a bug. %s\n", MACH0_(get_cputype_from_hdr)(&bin->hdr));
+		return false;
+	}
+
+	ut64 val = symbol_at;
+	if (reloc->pc_relative) {
+		val = symbol_at - pc - ins_len;
+	}
+
+	ut8 buf[8];
+	r_write_ble(buf, val, false, reloc->size * 8);
+	iob->write_at (iob->io, reloc->addr, buf, reloc->size);
+
+	return true;
+}
+
+static RList* patch_relocs(RBin *b) {
+	RList *ret = NULL;
+	RIO *io = NULL;
+	RBinObject *obj = NULL;
+	struct MACH0_(obj_t) *bin = NULL;
+	RIOMap *g = NULL, *s = NULL;
+	HtUU *relocs_by_sym = NULL;
+	SdbListIter *iter;
+	RBinInfo *info;
+	int cdsz;
+	ut64 n_vaddr, vaddr, size, offset = 0;
+	RIODesc *gotr2desc = NULL;
+
+	if (!b) {
+		return NULL;
+	}
+	io = b->iob.io;
+	if (!io || !io->desc)
+		return NULL;
+	obj = r_bin_cur_object (b);
+	if (!obj) {
+		return NULL;
+	}
+	bin = obj->bin_obj;
+
+	RSkipList * all_relocs = MACH0_(get_relocs)(bin);
+	if (!all_relocs) {
+		return NULL;
+	}
+	RList * ext_relocs = r_list_new ();
+	if (!ext_relocs) {
+		goto beach;
+	}
+	RSkipListNode *it;
+	struct reloc_t * reloc;
+	r_skiplist_foreach (all_relocs, it, reloc) {
+		if (!reloc->external) {
+			continue;
+		}
+		r_list_append (ext_relocs, reloc);
+	}
+	int num_ext_relocs = r_list_length (ext_relocs);
+	if (!num_ext_relocs) {
+		goto beach;
+	}
+
+	if (!io->cached) {
```

Comment:
does it make sense to have this check before all the allocations? If a user does not set `io.cache=true` he cannot patch anything, so it should be better to check this asap.

Reply:
The reason why the check is here is because these external relocs are pretty uncommon, so i want to avoid the user to get the warning all the time, and i moved it after we know if actually there are such type of relocs.

I can easily move it at the beginning, but then i'll have to remove the warning, which is not very user-friendly either.

Reply:
I see. In ELF we do have the check at the beginning, but if it is uncommon for MACH0 I'm fine keeping it here.